pageserver/src/tenant.rs

```diff
@@ -623,11 +651,40 @@ impl Tenant {
             "attach tenant",
             false,
             async move {
-                match tenant_clone.attach(ctx).await {
-                    Ok(_) => {}
+                let mut marker_file_presence = marker_file_presence;
+                match tenant_clone.attach(&mut marker_file_presence, ctx).await {
+                    Ok(_) => {
+                        info!(?marker_file_presence, "tenant attached");
+                        assert!(matches!(
+                            marker_file_presence,
+                            AttachMarkerFilePresence::Absent
+                        ));
+                    }
                     Err(e) => {
-                        tenant_clone.set_broken(e.to_string());
-                        error!("error attaching tenant: {:?}", e);
+                        error!(?marker_file_presence, "error attaching tenant: {:?}", e);
+                        match marker_file_presence {
+                            AttachMarkerFilePresence::Absent => {
+                                // We can reach here if we fail to create the marker file.
+                                //
+                                // The tenant may or may not be in the tenant map already,
+                                // it depends on our caller.
+                                //
+                                // There is no on-disk state for this tenant (empty tenant dir doesn't count.)
+                                //
+                                // It is crucial to not report the tenant attachment status as Attached.
+                                // So, we must not transition into Broken state.
+                                // It would be best to just delete the tenant object from the tenants
+                                // map, but, we're not in control of that here.
+                                todo!("find a solution")
```

Comment:
Ideas for this?

Reply:
I was thinking this would concerns the activation failed most likely, at least right now. The whole idea of this parameter is to carry status of cleanup activities. I wonder if it's interesting just what it ended up at, but also what it started as. It seems we are now just blinding our eyes for the 2x2 trying to process this as boolean.

Or is this a good indicator of progress at all, compared to having error types instead of anyhow? That might still be orthogonal to all this, but still I think it might make it easier to decide here.

Reply:
I don't know what to take from that comment. My interpretation:

1. You agree with the problem pointed out above the todo. I.e., that 
   `It is crucial to not report the tenant attachment status as Attached.`
    because
   `There is no on-disk state for this tenant (empty tenant dir doesn't count.)`
2. You are not proposing a solution to it.
3. But you are saying that you'd prefer to communicate the failure to create the marker file via a distinguished error, and arrive at the above `todo!` based on inspection of the error variant.

Ignore the below if the above is an incorrect interpretation.

Regarding (3), I think the presence status is better than an error variant because it allows for the assert in the `Ok()` case. But, I'm Ok to change to the error if that's what you prefer.
This is an unimportant detail to me.

What about (1) and (2) though?

Reply:
Obsolete, see https://github.com/neondatabase/neon/pull/4168#issuecomment-1540375991